File: DrakborgenDotNet/Network.cs
using Drakborgen;
using System.Net;
using System;
using SFML.Audio;
using static System.Runtime.InteropServices.JavaScript.JSType;
using static Drakborgen.Network;

namespace Drakborgen
{
    internal class Network : IDisposable
    {
        private bool _run = true;
        private Thread? _networkThread;
        private INetworkHost _networkHost;
        private bool _disposed = false;

        public bool IsConnected { get; private set; }

        public Network()
        {
            // Start the network thread
            /*_networkThread = new Thread(NetworkThreadProc)
            {
                IsBackground = true
            };
            _networkThread.Start();
            Console.WriteLine("Network thread started.");*/
        }

        ~Network()
        {
            Dispose(false); // Finalizer calls Dispose(false) for unmanaged cleanup
        }

        // IDisposable implementation
        void IDisposable.Dispose()
        {
            Dispose(true);
            GC.SuppressFinalize(this);
        }

        protected virtual void Dispose(bool disposing)
        {
            if (!_disposed)
            {
                Console.WriteLine("Network Dispose start");

                if (disposing)
                {
                    // Signal the thread to stop
                    _run = false;

                    // If the thread is running, wait for it to finish
                    if (_networkThread != null && _networkThread.IsAlive)
                    {
                        _networkThread.Join();
                        _networkThread = null;
                    }
                }

                Console.WriteLine("Network Dispose end");
                _disposed = true;
            }
        }

        private class TcpSocket
        {
            internal enum Status
            {
                Done
            }

            internal Status Receive(char[] data, int v, int received)
            {
                throw new NotImplementedException();
            }

            internal Status Connect(IpAddress ipAddress, int v)
            {
                throw new NotImplementedException();
            }

            internal Status Send(char[] data, int v, int sent)
            {
                throw new NotImplementedException();
            }
        }
        private class TcpListener
        {
            internal TcpSocket.Status Listen(int v)
            {
                throw new NotImplementedException();
            }

            internal TcpSocket.Status Accept(TcpSocket socket)
            {
                throw new NotImplementedException();
            }
        }
        private class IpAddress
        {
            private string _serverAddress;

            public IpAddress(string serverAddress)
            {
                _serverAddress = serverAddress;
            }
        }

        internal interface INetworkHost
        {
            void Run();
        }

        private class NetworkServer : INetworkHost
        {
            private record Client(TcpSocket Socket, bool IsConnected = false)
            {
                internal bool IsConnected { get; set; } = IsConnected;
            }

            Network _network;

            Client[] _clients;
            int[] _playerOrder;
            int _activePlayerIndex = 0;

            internal NetworkServer(Network network, int clientCount, Random random)
            {
                _network = network;
                _clients = new Client[clientCount];
                _playerOrder = new int[clientCount + 1];
                _playerOrder[0] = 0;
                for (int i = 0; i < clientCount; ++i)
                {
                    _clients[i] = new Client(new TcpSocket());
                    _playerOrder[i + 1] = i + 1; // Should index at i + 1, right?
                }
                random.Shuffle<int>(_playerOrder);
            }

            public void Run()
            {
                try
                {
                    while (_network._run)
                    {
                        if (!_network.IsConnected)
                        {
                            TcpListener listener = new TcpListener();

                            // bind the listener to a port
                            if (listener.Listen(53000) != TcpSocket.Status.Done)
                            {
                                throw new Exception();
                            }

                            // accept a new connection
                            for (int i = 0; i < _clients.Length; ++i)
                            {
                                Client client = _clients[i];

                                while (!client.IsConnected)
                                {
                                    Console.WriteLine("Connecting client " + i);

                                    TcpSocket.Status status = listener.Accept(client.Socket);

                                    Console.WriteLine("Client " + i + " connected with status " + Enum.GetName(status));

                                    client.IsConnected = status == TcpSocket.Status.Done;
                                }
                            }

                            _network.IsConnected = true;
                        }

                        while (_network.IsConnected)
                        {
                            Thread.Sleep(5000);

                            for (int i = 0; i < _clients.Length && _network.IsConnected; ++i)
                            {
                                Client client = _clients[i];

                                char[] data = new char[10];
                                int received = 0;
                                TcpSocket.Status status = client.Socket.Receive(data, 10, received);

                                Console.WriteLine("received " + received + " bytes from client " + i + " with status " + Enum.GetName(status) + ", " + data + " was the message");

                                client.IsConnected = status == TcpSocket.Status.Done;

                                if (!client.IsConnected)
                                {
                                    _network.IsConnected = false;
                                }
                            }
                        }
                    }
                }
                catch (Exception e)

                {
                    Console.Error.WriteLine(e.ToString());
                }
            }
        }

        private class NetworkClient : INetworkHost
        {
            private Network _network;
            private string _serverAddress;
            private TcpSocket _serverSocket;

            internal NetworkClient(Network network, string serverAddress)
            {
                _network = network;
                _serverAddress = serverAddress;
                _serverSocket = new TcpSocket();
            }

            public void Run()
            {
                try
                {
                    while (_network._run)
                    {
                        while (!_network.IsConnected)
                        {
                            Console.WriteLine("Connecting to server");

                            TcpSocket.Status status = _serverSocket.Connect(new IpAddress(_serverAddress), 53000);

                            Console.WriteLine("Server accepted with status " + Enum.GetName(status));

                            _network.IsConnected = status == TcpSocket.Status.Done;
                        }

                        while (_network.IsConnected)
                        {
                            Thread.Sleep(2000);

                            char[] data = ['g'];
                            int sent = 0;
                            TcpSocket.Status status = _serverSocket.Send(data, 1, sent);
                            Console.WriteLine("sent " + sent + " bytes with status " + Enum.GetName(status));

                            _network.IsConnected = status == TcpSocket.Status.Done;
                        }
                    }
                }
                catch (Exception e)

                {
                    Console.Error.WriteLine(e.ToString());
                }

            }
        }

        internal void StartServer(int clientCount, Random random)
        {
            IsConnected = false;
            _networkHost = new NetworkServer(this, clientCount, random);
            _networkThread = new Thread(() => _networkHost.Run());
        }

        internal void StartClient(string serverAddress)
        {
            IsConnected = false;
            _networkHost = new NetworkClient(this, serverAddress);
            _networkThread = new Thread(() => _networkHost.Run());
        }
    }
}

File: src/Drakborgen/Network.cpp
#include "Network.h"

#include "System.h"
#include "Random.h"

#include <SFML/Network.hpp>

#include <iostream>

#ifdef _WIN32
#include <windows.h>
#endif

class NetworkServer
{
public:
	NetworkServer(Network& network, int clientCount, Random& random)
		: network(network)
	{
		clients.resize(clientCount);
		playerOrder.resize(clientCount + 1);
		playerOrder[0] = 0;
		for (int i = 0; i < clientCount; ++i)
		{
			clients[i].socket = std::make_unique<sf::TcpSocket>();
			playerOrder[i + 1] = i + 1; // Should index at i + 1, right?
		}
		random.shuffle(playerOrder);
	}

	void operator()();

private:
	Network& network;

	struct Client
	{
		std::unique_ptr<sf::TcpSocket> socket;
		bool connected = false;
	};

	std::vector<Client> clients;
	std::vector<int> playerOrder;
	int activePlayerIndex = 0;
};

class NetworkClient
{
public:
	NetworkClient(Network& network, const std::string& serverAddress)
		: network(network)
		, serverAddress(serverAddress)
	{
		serverSocket = std::make_unique<sf::TcpSocket>();
	}

	void operator()();

private:
	Network& network;

	std::string serverAddress;

	std::unique_ptr<sf::TcpSocket> serverSocket;
};

void NetworkServer::operator()()
{
	try
	{
		while (network.run)
		{
			if (!network.connected)
			{
				sf::TcpListener listener;

				// bind the listener to a port
				if (listener.listen(53000) != sf::Socket::Status::Done)
				{
					THROW;
				}

				// accept a new connection
				for (int i = 0; i < clients.size(); ++i)
				{
					Client& client = clients[i];

					while (!client.connected)
					{
						std::cout << "Connecting client " << i << std::endl;

						sf::Socket::Status status = listener.accept(*client.socket);

						std::cout << "Client " << i << " connected with status ";

						switch (status)
						{
						case sf::Socket::Done:
							std::cout << "Done";
							break;
						case sf::Socket::NotReady:
							std::cout << "NotReady";
							break;
						case sf::Socket::Partial:
							std::cout << "Partial";
							break;
						case sf::Socket::Disconnected:
							std::cout << "Disconnected";
							break;
						case sf::Socket::Error:
							std::cout << "Error";
							break;
						default:
							break;
						}

						std::cout << std::endl;

						client.connected = status == sf::Socket::Done;
					}
				}

				network.connected = true;
			}

			while (network.connected)
			{
				Sleep(5000);

				for (int i = 0; i < clients.size() && network.connected; ++i)
				{
					Client& client = clients[i];

					char data[10];
					size_t received = 0;
					sf::Socket::Status status = client.socket->receive(data, 10, received);

					std::cout << "received " << received << " bytes from client " << i << " with status ";

					switch (status)
					{
					case sf::Socket::Done:
						std::cout << "Done, " << data[0] << " was the message";
						break;
					case sf::Socket::NotReady:
						std::cout << "NotReady";
						break;
					case sf::Socket::Partial:
						std::cout << "Partial";
						break;
					case sf::Socket::Disconnected:
						std::cout << "Disconnected";
						break;
					case sf::Socket::Error:
						std::cout << "Error";
						break;
					default:
						break;
					}

					std::cout << std::endl;

					client.connected = status == sf::Socket::Done;

					if (!client.connected)
					{
						network.connected = false;
					}
				}
			}
		}
	}
	catch (std::exception& e)
	{
		SetConsoleTextAttribute(GetStdHandle(STD_OUTPUT_HANDLE), WORD(FOREGROUND_RED | FOREGROUND_INTENSITY));
		std::cerr << std::endl << e.what() << std::endl;
		SetConsoleTextAttribute(GetStdHandle(STD_OUTPUT_HANDLE), WORD(FOREGROUND_RED | FOREGROUND_GREEN | FOREGROUND_BLUE | FOREGROUND_INTENSITY));
	}
}

void NetworkClient::operator()()
{
	try
	{
		while (network.run)
		{
			while (!network.connected)
			{
				std::cout << "Connecting to server" << std::endl;

				sf::Socket::Status status = serverSocket->connect(sf::IpAddress(serverAddress), 53000);

				std::cout << "Server accepted with status ";

				switch (status)
				{
				case sf::Socket::Done:
					std::cout << "Done";
					break;
				case sf::Socket::NotReady:
					std::cout << "NotReady";
					break;
				case sf::Socket::Partial:
					std::cout << "Partial";
					break;
				case sf::Socket::Disconnected:
					std::cout << "Disconnected";
					break;
				case sf::Socket::Error:
					std::cout << "Error";
					break;
				default:
					break;
				}

				std::cout << std::endl;

				network.connected = status == sf::Socket::Done;
			}

			while (network.connected)
			{
				Sleep(2000);

				const char data[] = { 'g' };
				size_t sent = 0;
				sf::Socket::Status status = serverSocket->send(data, 1, sent);
				std::cout << "sent " << sent << " bytes with status ";

				switch (status)
				{
				case sf::Socket::Done:
					std::cout << "Done";
					break;
				case sf::Socket::NotReady:
					std::cout << "NotReady";
					break;
				case sf::Socket::Partial:
					std::cout << "Partial";
					break;
				case sf::Socket::Disconnected:
					std::cout << "Disconnected";
					break;
				case sf::Socket::Error:
					std::cout << "Error";
					break;
				default:
					break;
				}

				std::cout << std::endl;

				network.connected = status == sf::Socket::Done;
			}
		}
	}
	catch (std::exception& e)
	{
		SetConsoleTextAttribute(GetStdHandle(STD_OUTPUT_HANDLE), WORD(FOREGROUND_RED | FOREGROUND_INTENSITY));
		std::cerr << std::endl << e.what() << std::endl;
		SetConsoleTextAttribute(GetStdHandle(STD_OUTPUT_HANDLE), WORD(FOREGROUND_RED | FOREGROUND_GREEN | FOREGROUND_BLUE | FOREGROUND_INTENSITY));
	}
}

void Network::startServer(int clientCount, Random& random)
{
	connected = false;
	networkThread = std::thread(NetworkServer(*this, clientCount, random));
	networkThreadStarted = true;
}

void Network::startClient(const std::string& serverAddress)
{
	connected = false;
	networkThread = std::thread(NetworkClient(*this, serverAddress));
	networkThreadStarted = true;
}
